core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java

```diff
@@ -25,14 +34,21 @@
    private static final int TOPOLOGY_ID_STOPPED = Integer.MAX_VALUE;
 
    private final ReadWriteLock ownershipLock = new ReentrantReadWriteLock();
+   private Executor remoteExecutor;
 
    private volatile int topologyId = -1;
    private final Lock topologyLock = new ReentrantLock();
-   private final Condition topologyCondition = topologyLock.newCondition();
+   private final Map<Integer, CompletableFuture<Void>> topologyFutures = new HashMap<>();
```

Comment:
Please wait for now, I started going it that way myself on my ISPN-5469 branch :)